### PR TITLE
cmd/evm: add b11r tool

### PR DESCRIPTION
**Pull request by lightclient**
Adds `b11r` tool to `evm`. This program can build and seal blocks from JSON input.

### CODE REVIEW

File: cmd/evm/internal/t8ntool/block.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"gopkg.in/urfave/cli.v1"
+)
+
+type blockInput struct {
+	Header *types.Header   `json:"header,omitempty"`
+	Uncles []*types.Header `json:"uncles,omitempty"`
+	TxRlp  string          `json:"txsRlp,omitempty"`
```
Comment (holiman) [inline cmd/evm/internal/t8ntool/block.go:35]: I think we should take txRoot, uncleRoot, not the full thing. 
Also, not sure about header, maybe we should take a custom env - preferrably on the same format as the env output from t8n, with some added fields

**Review by holiman — COMMENTED** [review]
Also - check out the `t8n_test`, it uses a pretty cool trick (not my invention) to re-execute itself as a command, and thus the test can execute the command itself and verify the expected output

**Comment by lightclient** [comment]
I think this is ready to review now. @holiman curious if this is still in line with what we discussed.

File: cmd/evm/internal/t8ntool/block.go [hunk 2]
```diff
@@ -0,0 +1,487 @@
...
+	err := rlp.DecodeBytes(raw, &txs)
+	if err != nil {
+		return nil, NewError(ErrorRlp, fmt.Errorf("unable to decode transaction from rlp data: %v", err))
+	}
+	inputData.Txs = txs
+
+	return inputData, nil
+}
+
+func readEnv(path string) (*bbEnv, error) {
```
Comment (holiman) [inline cmd/evm/internal/t8ntool/block.go:362]: These functions are very much repeated. Maybe we can make a utility method that we reuse: 
```golang

func readFile(path, desc string, dest interface{}) error{
       ....
     if err := decoder.Decode(dest); err != nil {
         return nil, NewError(ErrorJson, fmt.Errorf("failed unmarshaling %v from stdin: %v", desc, err))
     ....
}
```
And then just 
```
  env :=  &bbEnv
  err := readFile(path, "env", env)
```

Comment (lightclient) [reply cmd/evm/internal/t8ntool/block.go]: Great point, I've updated!

**Review by 09306677806 — COMMENTED** [review]
Can someone help me tell me how can I get my Atrium line and my website back?

**Comment by lightclient** [comment]
@09306677806 I recommend turning it off and then back on again. Studies have show it solves 98.7% of technical issues.

**Review by holiman — APPROVED** [review]
LGTM

**Comment by holiman** [comment]
@lightclient does my changes LGTY? Do you consider this done, or is there anything else to add/change?

**Comment by lightclient** [comment]
@holiman yes, your change look good to me. I'm currently using `b11r` to generate tests and it appears to working as expected. Over the next couple weeks, I'll be using it more thoroughly so some issues may be uncovered. Generally, I think LGTM since no one else depends on it and it is in a pretty good place at the moment.

File: cmd/evm/internal/t8ntool/block.go [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+	if i.Clique.Authorized != nil {
+		if i.Header.Nonce != nil {
+			return nil, NewError(ErrorConfig, fmt.Errorf("sealing with clique and voting will overwrite provided nonce"))
+		}
+		if *i.Clique.Authorized {
+			header.Nonce = [8]byte{}
+		} else {
+			header.Nonce = [8]byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
+		}
+	}
```
Comment (holiman) [inline cmd/evm/internal/t8ntool/block.go:221]: Not sure about this. `0xff...` is `nonceAuthVote`, and `0x00..` is `nonceDropVote`. I don't understand why you are using them in this way, setting it to one or the other depending on whether the sealer is authorized?

**Review by holiman — COMMENTED** [review]
One remaining thing I'm not sure about, otherwise LGTM

File: cmd/evm/internal/t8ntool/block.go (hunk 3 above)
Comment (lightclient) [reply cmd/evm/internal/t8ntool/block.go:221]: My goal was to encapsulate as much of the Clique logic from `b11r` caller as possible. So in this case, a user may want to test a chain Clique blocks where they vote in or out (auth or deauth) certain sealers. If the `Clique` object didn't have the `Authorized` field I suppose a user could hard code these values into the nonce. I don't feel strongly either way.

I realize now also that `Authorized` may be more confusing than `Authorize`, which is the term used in the EIP-225 tests section.

Comment (holiman) [reply cmd/evm/internal/t8ntool/block.go:221]: Ah.. so `Authorized` means `Make this block vote yes (to some proposal)` ?

Comment (lightclient) [reply cmd/evm/internal/t8ntool/block.go]: Thanks - it's been update.

**Pull request merged**